cmd/clusterctl/pkg/client/cluster/move.go

```diff
@@ -202,11 +232,412 @@ func (o *objectGraph) Discovery(namespace string, types []metav1.TypeMeta) error
+func (o *objectGraph) createTreeIntoTarget(ownerNode, nodeToCreate *node) error {
+
+	// Creates the Kubernetes object corresponding to the nodeToCreate.
+	// Nb. The operation is wrapped in a retry loop to make move more resilient to unexpected conditions.
+	err := retry(retryCreateTargetObject, retryIntervalCreateTargetObject, o.log, func() error {
+		return o.createTargetObject(ownerNode, nodeToCreate)
+	})
+	if err != nil {
+		return err
+	}
+
+	// Creates - in parallel - the dependents nodes and the softDependents nodes (with the respective object tree).
+	// The approach to concurrency here comes from golang.org/x/sync/errgroup.
+	var wg sync.WaitGroup
+	var gerr error
+	var errOnce sync.Once
+
+	for dependent := range nodeToCreate.dependents {
+		wg.Add(1)
+		go func(dependent *node) {
+			defer wg.Done()
+			if err := o.createTreeIntoTarget(nodeToCreate, dependent); err != nil {
+				errOnce.Do(func() { gerr = err })
+			}
+		}(dependent)
+	}
+
+	for dependent := range nodeToCreate.softDependents {
+		wg.Add(1)
+		go func(dependent *node) {
+			defer wg.Done()
+			if err := o.createTreeIntoTarget(nodeToCreate, dependent); err != nil {
+				errOnce.Do(func() { gerr = err })
+			}
+		}(dependent)
+	}
+
+	wg.Wait()
+	if gerr != nil {
+		return gerr
+	}
+
+	// If the nodeToCreate is a machine, wait for the machine to be reconciled and linked with the corresponding node.
+	if nodeToCreate.identity.APIVersion == clusterv1.GroupVersion.String() && nodeToCreate.identity.Kind == "Machine" {
+		o.log.V(1).Info("Waiting for reconciliation with the workload cluster node", "Machine", nodeToCreate.identity.Name, "Namespace", nodeToCreate.identity.Namespace)
+		machineKey := client.ObjectKey{Namespace: nodeToCreate.identity.Namespace, Name: nodeToCreate.identity.Name}
+		if err := o.machineWaiter(o.toProxy, machineKey); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// createTargetObject creates the Kubernetes object corresponding to the node in the target Management cluster, taking care of restoring the OwnerReference with the owner node, if any.
+func (o *objectGraph) createTargetObject(ownerNode, nodeToCreate *node) error {
+	o.log.V(1).Info("Creating", nodeToCreate.identity.Kind, nodeToCreate.identity.Name, "Namespace", nodeToCreate.identity.Namespace)
+
+	cFrom, err := o.fromProxy.NewClient()
+	if err != nil {
+		return err
+	}
+
+	// Get the source object
+	sourceObj := &unstructured.Unstructured{}
+	sourceObj.SetAPIVersion(nodeToCreate.identity.APIVersion)
+	sourceObj.SetKind(nodeToCreate.identity.Kind)
+	targetObjKey := client.ObjectKey{
+		Namespace: nodeToCreate.identity.Namespace,
+		Name:      nodeToCreate.identity.Name,
+	}
+
+	if err := cFrom.Get(ctx, targetObjKey, sourceObj); err != nil {
+		return errors.Wrapf(err, "error reading %q %s/%s",
+			sourceObj.GroupVersionKind(), sourceObj.GetNamespace(), sourceObj.GetName())
+	}
+
+	targetObj := sourceObj.DeepCopy()
```

Comment:
I'm not sure we need to clone the object - could we just modify sourceObj?

Reply:
done